hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/handler/CloseRegionHandler.java

```diff
@@ -52,85 +54,101 @@
-        // the master to split our logs in order to recover the data.
-        server.abort("Unrecoverable exception while closing region " +
-          regionInfo.getRegionNameAsString() + ", still finishing close", ioe);
-        throw new RuntimeException(ioe);
+      if (region.getCoprocessorHost() != null) {
+        // XXX: The behavior is a bit broken. At master side there is no FAILED_CLOSE state, so if
+        // there are exceptions thrown from the CP, we can not report the error to the master, and
+        // if here is, we just return without calling reportRegionStateTransition, the TRSP at
+        // master side will hang there for ever. So here if the CP throws an exception out, the only
+        // way is to abort the RS...
+        region.getCoprocessorHost().preClose(abort);
       }
-
-      this.rsServices.removeRegion(region, destination);
-      rsServices.reportRegionStateTransition(new RegionStateTransitionContext(TransitionCode.CLOSED,
-        HConstants.NO_SEQNUM, Procedure.NO_PROC_ID, -1, regionInfo));
-
+      if (region.close(abort) == null) {
+        // This should not happen. Left over from old days when RegionServer
+        // ran the splits? Now splits are run by the Master.
+        LOG.warn("Can't close {}, already closed; split? (Unexpected state - investigate)",
+          this.regionInfo.getRegionNameAsString());
+        // Return null here. Makes more sense. We didnt do close on the Region.
+        return;
+      }
+      this.rss.removeRegion(region, destination);
+      boolean b = reportRegionStateTransition();
       // Done!  Region is closed on this RS
-      LOG.debug("Closed " + region.getRegionInfo().getRegionNameAsString());
+      LOG.debug("Closed {} (notified master={})", this.regionInfo.getRegionNameAsString(), b);
+      // We don't care if we are able to inform the Master successfully; try but ignore response.
     } finally {
-      this.rsServices.getRegionsInTransitionInRS().
-        remove(this.regionInfo.getEncodedNameAsBytes(), Boolean.FALSE);
+      // Presumption is that before the Close was run, the CLOSE operation had been registered
+      // with rss.getRegionsInTransitionInRS(). Here we do cleanup whether success or failure.
+      this.rss.getRegionsInTransitionInRS().remove(this.regionInfo.getEncodedNameAsBytes(),
+        Boolean.FALSE);
     }
   }
+
+  /**
+   * @throws IOException Let subclasses throw IOE if they need to.
+   */
+  boolean reportRegionStateTransition() throws IOException {
+    return this.rss.reportRegionStateTransition(
+      new RegionStateTransitionContext(TransitionCode.CLOSED, HConstants.NO_SEQNUM,
+        getClosePID(), -1, this.regionInfo));
+  }
+
+  /**
+   * @return PID of Close to report to Master; default is Procedure.NO_PROC_ID.
+   */
+  long getClosePID() {
+    return Procedure.NO_PROC_ID;
+  }
+
+  @Override protected void handleException(Throwable t) {
+    LOG.warn("Fatal error occurred while closing region {}, aborting...",
```

Comment:
base class is slightly more careful/different in handling the abort. Is it better to just customize the log here and the call the `super` method to handle abort? or maybe the `super` implementation need to abort more aggressively (abort on all exceptions instead of just the types it does)?